src/coreComponents/dataRepository/Group.hpp

```diff
@@ -1329,6 +1330,19 @@ class Group
   integer getLogLevel() const { return m_logLevel; }
   ///@}
 
+  /**
+   * @brief Performs re-initialization of certain variable depending on the solver being used.
+   */
+  virtual void reinit() {}
+
+  /**
+   * @brief Return PyGroup type.
+   * @return Return PyGroup type.
+   */
+#if defined(GEOSX_USE_PYGEOSX)
+  virtual PyTypeObject * getPythonType() const;
+#endif
```

Comment:
So I'm definitely late about this PR, but why do we add a `python` dependency here in `Group`?
From there, "_everything_" in `geosx` will depend on `python`.

I know that "_everything_" being a `Group`, it's convenient that `python` gets access to `Groups`.
- But it will eventually break encapsulation. And it's `geosx`'s internal implementation; I doubt a wrapper should access it.
- We should be able to compile `geosx`, with no `python` dependency, even if `ENABLE_PYGEOSX` is `ON`. This `Group::getPythonType` just kills it.

Reply:
@sframba

Reply:
>
> why do we add a python dependency here in Group
> 

It's already included from `View.hpp` https://github.com/GEOSX/GEOSX/blob/01f4b624504ec0ee915a6f457ef8cc09d054b168/src/coreComponents/dataRepository/wrapperHelpers.hpp#L37

But also if you actually follow the includes I'm pretty sure that we don't actually `#include <python.h>` in any headers. I think we got away with forward declarations only.

https://github.com/GEOSX/LvArray/blob/ab0d3025d2962aad10457b7bf7f5b6d266232276/src/python/pythonForwardDeclarations.hpp

https://github.com/GEOSX/LvArray/search?l=C%2B%2B&p=3&q=python.h

> But it will eventually break encapsulation. And it's `geosx`'s internal implementation; I doubt an external 
wrapper should access it.

I disagree. We need to be able to access and modify arbitrary data from Python doing it without exposing `Group` and `Wrapper` sounds convoluted. Not to mention that the data repository isn't really an implementation detail, it's mirrored in the XML files we currently use.


You could imagine a solution where instead of putting virtual functions in `Group` and `Wrapper` we had a massive type switching block inside `pygeosx` that would get the `typeid` of the object and then downcast and export it to Python. But I've been in favor of removing that code pattern from GEOSX for a long time.

Reply:
> But also if you actually follow the includes I'm pretty sure that we don't actually `#include <python.h>` in any headers. I think we got away with forward declarations only.

I’m not sure to understand entirely. 
If we need to forward declare, it’s because we need something.
So we depend on this something.
In this case we’ll depend on `PyTypeObject` which will be found in `python` (most probably during link edition?), (unless the name is misleading).

> I disagree. We need to be able to access and modify arbitrary data from Python doing it without exposing `Group` and `Wrapper` sounds convoluted. 

It depends what you call convoluted. This surely is a matter of costs and benefits w.r.t. the global goal.

> Not to mention that the data repository isn't really an implementation detail, it's mirrored in the XML files we currently use.

Yes, I know.
- But for the end users, the software architecture should be an implementation detail. The input file should make their life easy, and there is not strong reason that a C++ hierarchy is the most convenient for them to fill.
- And as a developer, the fact that I’m not able to _reorganize_ my code freely without potentially impacting the end user is not the most comfortable situation. To my eyes, this breaks encapsulation.

> You could imagine a solution where instead of putting virtual functions in Group and Wrapper we had a massive type switching block inside pygeosx that would get the typeid of the object and then downcast and export it to Python.

It could be a way to invert dependencies, which can be helpful. So while apparently tedious, maybe this kind of pattern could help us for tricky situations?

Reply:
> I’m not sure to understand entirely.
> If we need to forward declare, it’s because we need something.
> So we depend on this something.
> In this case we’ll depend on `PyTypeObject` which will be found in `python` (most probably during link edition?), (unless the name is misleading).

Yes it requires that we link the main geosx library against Python, but it does not require that we include the python headers. This seems like a pretty big difference to me. Take for example conduit, which is something that is directly included in Wrapper to make writing restarts possible. This is directly analogous to Python in that it exports data stored in any Wrapper to another environment. Except that for conduit we need not only need to link but compile against the headers as well. As far as I can tell this hasn't produced any problems for GEOSX. Conduit is a third party library and if all of GEOSX depends on it then so be it. It's not even a question of compile time either because 99% of the time is spent compiling the CUDA kernels.

> It could be a way to invert dependencies, which can be helpful. So while apparently tedious, maybe this kind of pattern could help us for tricky situations?

This pattern is a pain in the ass to work with. It makes already tedious compilation errors 40 times harder to deal with, it leads to code that is hard to read, it is hard to maintain and extend. I also don't think this situation is particularly tricky. When building pygeosx we add Python as another top level dependency. If you don't want this then don't build with Python.

Reply:
> Yes it requires that we link the main geosx library against Python, but it does not require that we include the python headers. This seems like a pretty big difference to me.

Can you elaborate on this difference?
I’m more on the “_you need it, whatever the form_“ side. 

> Take for example conduit, which is something that is directly included in Wrapper to make writing restarts possible. This is directly analogous to Python in that it exports data stored in any Wrapper to another environment.

`conduit` looks mainly like a smart serializer/deserializer. It helps at exchanging data.
Using `python`, we could exchange processes and implementations too.

> As far as I can tell this hasn't produced any problems for GEOSX.

Actually, having those implementations exposed in the heart of the whole code base indeed produces problems.
Because it binds our questions to one unique solution.
It is imposed as the source of information, so it’s impossible to get a different one.

An `Event`, a `MeshGenerator` or a customized `Output` could be implemented in another language (e.g. `python`) because it could be more convenient.
It’s not possible because we have an implementation centric architecture.

And therefore whatever I want to do in GEOSX I have to deal with `conduit` of `xml` or `...`, even if that should be out of scope.
How can I mock the inputs of my `VTK` code and validate that this key feature is rock solid? I cannot, so there will be more bugs.

Reply:
> Can you elaborate on this difference?
> I’m more on the “_you need it, whatever the form_“ side.

If you try and call some Python function in a physics solver it will produce a compilation error because the header wasn't included. And your IDE doesn't get clogged up.

> Actually, having those implementations exposed in the heart of the whole code base indeed produces problems.
> Because it binds our questions to one unique solution.
> It is imposed as the source of information, so it’s impossible to get a different one.

How does optionally linking to Python impose anything as a source for information?

>An `Event`, a `MeshGenerator` or a customized `Output` could be implemented in another language (e.g. `python`)  because it could be more convenient.
> It’s not possible because we have an implementation centric architecture.

Yes one can infer from the dependencies and source of data repository that we write our restarts with Conduit. But I don't see how that would make it any harder to create an alternate restart implementation that used something else entirely or was perhaps in a different language.

> And therefore whatever I want to do in GEOSX I have to deal with conduit of xml or ..., even if that should be out of scope.

This sounds like it would involve changes way beyond the scope of this PR, basically a wholesale rewrite of the data repository. Again though I don't see the value. I can't speak for others, but when I'm working on something for a physics solver, I never even think about Conduit or any of the other TPL's that I'm not actively using.

> How can I mock the inputs of my VTK code and validate that this key feature is rock solid? I cannot, so there will be more bugs.

Can you be specific as to what it is that you'd like to do, and how having GEOSX link to Python prevents you from doing it?

Reply:
Hi @corbett5 sorry for the delay and thank you for the fruitful discussion, I appreciate! 😉 

> If you try and call some Python function in a physics solver it will produce a compilation error because the header wasn't included. And your IDE doesn't get clogged up.

So in a nutshell, you trim to the maximum what can and cannot be used.
If anyone wants more, they will have to explicitly add the `include` (or the forward declaration) to make it work.
This makes sense, meanwhile I personally prefer to be extremely strict in the build: when a dependency is added to the `CMakeLists.txt` it's really easy to detect.

For the rest of the discussion, let's consider the following simplified example.
I have a `Solver` which generates a `ComplicatedOutput` that I want to `WriteOnDisk`.
This operation being critical for my users, I want to perform extensive testing.
Today in `GEOSX` I may have to create a lot of simulations, such that all my `Solvers` will be able to create a lot of different `ComplicatedOutput` and I'll test `WriteOnDisk` with all these data coming from real workflows.
I have to abide by the `GEOSX` organization (`xml`, `conduit`, running the simulation on multiple ranks, etc.).

But instead, I'd like to be able to implement "from scratch" my own versions of `ComplicatedOutput` dedicated to testing.
I could be able to generate batches of cases without running any simulation (therefore in the unit tests and not in the integrated tests).
I could even be able to generate "_wrong_" data or dangerous corner cases, simulate a `NaN` quite conveniently and see how my `WriteOnDisk` behaves.

This is what I mean by "_different source of data_": I'm bound to the current organization and implementation and it's complicated to escape. It's not `xml` or `conduit` themselves, but more about our implementation centric architecture.

> This sounds like it would involve changes way beyond the scope of this PR

Absolutely, this is where the discussion takes place, but not where actions (if any) would happen.

> I can't speak for others, but when I'm working on something for a physics solver, I never even think about Conduit or any of the other TPL's that I'm not actively using.

Absolutely, when you are working in `GEOSX` _à la_ `GEOSX`, there is no limitation and things flow seamlessly.
But things may be a little different when you try to plug other components for whatever reason (I mentioned testing _e.g._).

Reply:
> I could be able to generate batches of cases without running any simulation (therefore in the unit tests and not in the integrated tests)

I certainly agree this would be super useful and God knows we're really lacking on the unit test front for a lot of our capabilities. But I don't really see how depending on `conduit` or `xml` or `Python` really complicates things here. We should have some generic unit test helpers that can create a bunch of bare bones GEOSX data structures (basically a mesh) and then it would be easy enough for your unit tests to create and populate some dummy fields that could then be written out. Some of the physics solver tests basically do this. Or if you write `WriteOnDisk` in a modular enough way then maybe you could just manufacture the test data without the Data Repository.

Reply:
> But I don't really see how depending on conduit or xml or Python really complicates things here. We should have some generic unit test helpers that can create a bunch of bare bones GEOSX data structures (basically a mesh) and then it would be easy enough for your unit tests to create and populate some dummy fields that could then be written out.

The main idea I have in mind, in a nutshell, is that because of the current construction (not specially `xml` or `conduit`, it's a bit more general), one surely has to implement more than the shallow layer that could be required.

Like you said, if I want to test my `WriteOnDisk` for all kinds of inputs, I'd like to be able to do so without having to create any of the `GEOSX` infrastructure: I just want a `ComplicatedOutput`. If I just want the skins, I do not want the bare bones! 😉 And today in `GEOSX` this is not straightforward.
Consider the `testTriaxial` example. This test looks simple, but we nevertheless have to deal with `FunctionManager`, `ConstitutiveManager`, `xml` input file, `Group` and `conduit::node`, `GeosxState`. It's more a mini `GEOSX` simulation than a unit test: it's complicated to test at the "_module_" level.

I'll be in LLNL on a regular basis now, so I'll hope we can find some time to interact on concrete cases, that will be more efficient!

Reply:
Yeah I'd be down to talk about what we could do to make unit testing more accessible. Unfortunately though I'm now living in New Mexico so...

Reply:
I'm working on "propositions" about this, I'll be doing a presentation with (hopefully) concrete examples.
It's with great pleasure that I would get you involved!